### PR TITLE
ENH: integrate: quad() for vector-valued functions

### PR DESCRIPTION
**Pull request by pv**
Implement a variant of the QUADPACK integration algorithm in Python, and
use it for integrating vector-valued functions.

This is useful in cases in which the function is not easy to calculate
elementwise, or a different norm than the max-norm is wanted for error
estimation, in which cases running quad() separately for each element is
not optimal.

The vectorized algorithm also allows for some degree of parallelization.

Example case:
```python
import time
import numpy as np

from scipy.integrate import quad, solve_ivp

t = np.linspace(0, 10, 200)

def f(x):
    sol = solve_ivp(lambda t, y: [x - y[1], x + y[0]], (0, 10), [1, 0], t_eval=t)
    return sol.y[0]

start = time.monotonic()
res = quad(f, -1, 1, vectorized=True, epsrel=1e-3)[0]
print("vec:", time.monotonic() - start)

start = time.monotonic()
res2 = np.array([quad(lambda x: f(x)[j], -1, 1, epsrel=1e-3)[0] for j in range(len(t))])
print("novec:", time.monotonic() - start)

import matplotlib.pyplot as plt
plt.plot(t, res, '-', t, res2, '--')
plt.show()
```
The vectorized=True scales linearly with `len(t)`, vectorized=False quadratically.
For scalar problems, there's more Python overhead:
```
In [5]: %timeit quad(np.sin, 0, 3, vectorized=True)
212 µs ± 48.5 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
In [6]: %timeit quad(np.sin, 0, 3, vectorized=False)
17.3 µs ± 2.21 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
On pypy, it goes the other way for pure-Python code:
```
[PyPy 6.0.0 with GCC 8.0.1 20180324 (Red Hat 8.0.1-0.20)] on linux
In [5]: %timeit quad(lambda x: 1/(1 + x**2), 0, 3, vectorized=True)                            
43 µs ± 3.49 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
In [6]: %timeit quad(lambda x: 1/(1 + x**2), 0, 3, vectorized=False)                           
103 µs ± 5.26 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Evaluation counts are similar in most cases:
```
In [20]: quad(lambda x: 1/(1 + x**2), 0, 3, vectorized=False, full_output=True)[2]['neval']    
63
In [21]: quad(lambda x: 1/(1 + x**2), 0, 3, vectorized=True, full_output=True)[2]['neval']     
63
```

### CODE REVIEW

File: scipy/integrate/_quad_vec.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+    eps = sys.float_info.epsilon
+    round_err = float(norm_func(50 * eps * h * s_k_abs))
+
+    if round_err > sys.float_info.min:
+        err = max(err, round_err)
+
+    return h * s_k, err, round_err
+
+
+_quadrature_gk21.num_eval = 21
```
Comment (tylerjereddy) [inline scipy/integrate/_quad_vec.py]: Is this function attribute just a form of documentation? `range(21)` is used twice in the function above and there's also a product using `21`--is there no sensible way to abstract the magic number `21` to an attribute / variable beforehand?

Are there cases where `num_eval` for a gk21 function is `!= 21`?

Comment (pv) [reply scipy/integrate/_quad_vec.py]: No, it's used to count number of evaluations. It's 21-point Gauss-Kronrod rule. Computing the points is not completely straightforward, so it needs to be hardcoded.

**Comment by ev-br** [comment]
QUADPACK seems to use 15-point quadratures for semi-infinite and infinite intervals (dqagie.f), instead of 21-point rules, do you know if there's a reason for it?

**Comment by ev-br** [comment]
Had a quick look into the QUADPACK book:

> In GAQI rules of lower degree are used... since it is likely, due to to the transformation that a singularity occurs in the transformed integrand at the point 0, so little would be gained from using a formula of a higher degree.

**Comment by ev-br** [comment]
An immediate question is about the UX: I'm not a big fan of `vectorized` kwargs, and more generally, of interdependent arguments which make sense only for some combinatons of other parameters. 

Here, vectorized and "classic" codepaths seem to have enough differences to warrant a public `quad_vec` function, I'd think.

**Comment by pv** [comment]
Right. I put it currently as vectorized= kwarg as it seemed the code paths are more similar than different, but it is true that how the arguments work gets more complicated.

**Comment by pv** [comment]
Removed any changes to `quad()`, making this a completely separate addition.
Use now the lower-order quadrature for infinite intervals.

**Comment by ev-br** [comment]
+1 from me. 
TravisCI failure is/was a trivial pep8 issue, pypy failure can be ignored, I think.
This PR addresses a very common issue/feature request (a bunch of questions on SO, https://stackoverflow.com/questions/41223186/numpy-vectorization-with-integration and many others). 
I'd like to merge this early in the release cycle, and iterate if needed. 
Thoughts?

**Comment by rgommers** [comment]
> I'd like to merge this early in the release cycle, and iterate if needed.
Thoughts?

Makes sense, go for it I'd say. Looks like a nice enhancement from a quick read.

**Comment by ev-br** [comment]
Ok, in it goes. Thanks Pauli, all.
 This should be among the highlights so a release note would be added at some point.

File: scipy/integrate/quadpack.py [hunk 2]
```diff
@@ -333,15 +386,39 @@ def quad(func, a, b, args=(), full_output=0, epsabs=1.49e-8, epsrel=1.49e-8,
...
+                           points)
+        else:
+            retval = _quad_weight(func, a, b, args, full_output, epsabs, epsrel,
+                                  limlst, limit, maxp1, weight, wvar, wopts)
     else:
-        retval = _quad_weight(func, a, b, args, full_output, epsabs, epsrel,
-                              limlst, limit, maxp1, weight, wvar, wopts)
+        if weight is not None:
+            raise ValueError("Option 'weight' not supported when vectorized=True")
+        if args:
```
Comment (AnirudhDagar) [inline scipy/integrate/quadpack.py]: @tupui this was the initially proposed behaviour I was suggesting in #14426

File: scipy/integrate/quadpack.py [hunk 3]
```diff
@@ -386,39 +333,15 @@ def quad(func, a, b, args=(), full_output=0, epsabs=1.49e-8, epsrel=1.49e-8,
...
-        else:
-            retval = _quad_weight(func, a, b, args, full_output, epsabs, epsrel,
-                                  limlst, limit, maxp1, weight, wvar, wopts)
+    if weight is None:
+        retval = _quad(func, a, b, args, full_output, epsabs, epsrel, limit,
+                       points)
     else:
-        if weight is not None:
-            raise ValueError("Option 'weight' not supported when vectorized=True")
-        if args:
```
Comment (AnirudhDagar) [inline scipy/integrate/quadpack.py]: @tupui the args functionality was removed here but not added to the new function `quad_vec`.

**Pull request merged**